index.rst

```diff
@@ -51,11 +51,182 @@
+A temporary caching datastore would not have this problem and may be the short term compromise.
+
+Client/Server Butler
+--------------------
+
+Another option is to change the ``Butler`` constructor to be a factory method that can return a classical ``Butler`` or a ``RemoteButler``.
+This remote butler would implement the get, put, ingest, and prune methods directly as calls to the remote server.
+It would be easier to handle transactions in this context because all of the required work that requires the transactions would be handled by the server since the APIs are at a much higher level than registry.
+
+Implementing a client/server ``Butler.put()`` is difficult because there is no way to convert the Python in-memory dataset to serialized form without involving the formatter infrastructure that is called within ``Datastore``.
+One option is for the server implementation of ``Butler.put()`` to not take the dataset at all, but to instead store a placeholder entry in the datastore registry and then return a signed URL, along with the ``DatasetRef``, that can be used by the client code to push the file.
+The client code would then use a local Datastore implementation to create the file from the relevant formatter and then upload it.
+The downside of this approach is that it is not immediately clear how to handle composite disassembly (where the file is split into multiple components and each is stored separately in the file store) since that is a datastore configuration.
+
+A ``Butler.get()`` would most logically be handled as a call to ``Butler.getURIs()`` (probably via the datastore ``getStoredItemsInfo`` method) to obtain the signed URL, download the file, and then use a local Datastore implementation to read that file.
+This could conceivably leverage the purported "caching datastore" concept.
+
+Even if all this is made to work, users still expect to be able to use many of the registry methods for querying the collections, datasets and dimension records.
+This suggests that it might be better to implement the client/server registry first and build on that, and then subsequently add explicit ``Butler`` overrides if performance is an issue (for example if ``Butler`` calls registry methods in a loop there may be significant run time overheads).
+
+Manager Classes
+---------------
+
+The final option is to implement each of the manager classes as client/server implementations.
+The registry is already designed to support pluggable managers and they are already designed to isolate database access.
+This seems like the cleanest way forward but the interfaces are very low level and this makes it significantly harder for non-Butler clients to do anything useful with the data being returned.
+It also raises the possibility of the interfaces being very slow when called in loops and may require significant caching of results and also the addition of new methods that move loops into the server.
```

Comment:
I don't think the last sentence is a concern; the `Manager` interfaces are pretty thoroughly vectorized already.

But overall, I think I now share your pessimism that client/server managers will work well; there's too much SQLAlchemy in those interfaces.